### PR TITLE
Claude caching at depth

### PR DESCRIPTION
**Pull request by ghost**
Very optional, very out of the way, very explicit warnings about how badly it can go, ample room for comfortable configuration ({{random}} on prefills should work out of the box, which is the most useful usecase for {{random}}, PHIs and A/Ns take minor configuration).

## Checklist:

- [X] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

### CODE REVIEW

File: default/config.yaml [hunk 1]
```diff
@@ -168,5 +168,12 @@ claude:
   # (e.g {{random}} macro or lorebooks not as in-chat injections).
   # Otherwise, you'll just waste money on cache misses.
   enableSystemPromptCache: false
+  # Enables caching of the message history at depth (if supported).
+  # https://docs.anthropic.com/en/docs/build-with-claude/prompt-caching
+  # -- IMPORTANT! --
+  # Use with caution. Behavior may be unpredictable and no guarantees can or will be made.
+  # Set to an integer to specify the desired depth. 0 (which does NOT include the prefill)
+  # should be ideal for most use cases.
+  cachingAtDepth: false
```
Comment (Cohee1207) [inline default/config.yaml]: Some naming consistence would be nice.
`enableCachingAtDepth`?

Comment (ghost) [reply default/config.yaml]: The config value is an integer and not a boolean. The boolean `false` just disables the feature.

Comment (ghost) [reply default/config.yaml]: To elaborate, it pretty much _has_ to be a configurable integer so you can move the caching behind your PHI/depth injections and avoid just caching nonsense on every prompt. An extension of why I ignore the prefill but more generic and less consistently repeatable.

Comment (Cohee1207) [reply default/config.yaml]: I think the disabled value should be not boolean then.

Comment (Cohee1207) [reply default/config.yaml]: i.e. `-1`

Comment (ghost) [reply default/config.yaml]: Sure, I'll commit that change in. By the by, openrouter also supports claude caching but I can't find the request code at all to add the `cache_control` field in, can you point it out for me?

Comment (Cohee1207) [reply default/config.yaml]: Put it somewhere around here:

https://github.com/SillyTavern/SillyTavern/blob/a3ca407b2714df5af5a9f83aa925fd64fb778e24/src/endpoints/backends/chat-completions.js#L843

**Comment by ghost** [comment]
OpenRouter support added as well. To clarify, the 2 in the code is supposed to be from the cache immediately before the previous message (if available) regardless of what's going on with injections at depth (which should all presumably have user role for Claude?), hence why _depth_ is strict role switches (unless I'm doing something absurdly stupid here, always possible).

**Comment by cloak1505** [comment]
> (which should all presumably have user role for Claude?)

I doubt that's a requirement. It's on user messages in the docs because naturally the last message in a request would just be user (solo chat being the typical use case). User sends on turn 1. Then turn 2. The current and previous breakpoints happen to be on the user; you get the idea.

> hence why _depth_ is strict role switches (unless I'm doing something absurdly stupid here, always possible)

https://docs.anthropic.com/en/release-notes/api#october-8th-2024
Anthropic recently loosened restrictions to allow for consecutive same-role messages, but they say those messages will be combined into single message, which makes me worry whether consecutive assistant messages will break the cache. If it does, then we'll have to default to caching only the user messages to be safe. If not, well, caching assistant messages will help group chats where most messages are from various characters.

Also, we are allowed up to 4 breakpoints, so we should use 4 (edit: if not `enableSystemPromptCache` else ~~3~~ uh, 2). This will allow users to edit 7 messages back (after 4th last user turn) and keep the cache.

#2693
> My original idea was system prompt + last 3 user messages, which would allow you to restart the chat assuming the sys prompt is at least 1024 tokens

**Comment by ghost** [comment]
I very intentionally only used 2 breakpoints to avoid breaking the system prompt caching option (which uses up to 2 breakpoints).

EDIT: Anyway the current setup is entirely rational. It's customizable, it hits good caches between swipes and between messages, assuming people aren't going back and editing the chat history. Extremely optimized for 90% of usecases _other than_ swiping 50 times (but still optimized for that).

EDIT 2: Exclusively caching on user messages might be a good heuristic, I'll sleep on it. Group chats were always a mess and will remain a mess and I refuse to waste more than 5 neurons on them.

**Comment by cloak1505** [comment]
System prompt caching uses 2? My bad. But where and why?

**Comment by ghost** [comment]
https://github.com/SillyTavern/SillyTavern/blob/a3ca407b2714df5af5a9f83aa925fd64fb778e24/src/endpoints/backends/chat-completions.js#L117
https://github.com/SillyTavern/SillyTavern/blob/a3ca407b2714df5af5a9f83aa925fd64fb778e24/src/endpoints/backends/chat-completions.js#L137

One breakpoint on the sysprompt and another for the tools. It doesn't NEED both, but if I were using a third breakpoint I'd just place it on the prefill to optimize swipes all the way (with a third configuration option).

EDIT: To clarify the prefill breakpoint would be more of a "might as well" thing because your prefill shouldn't be big enough for the caching to make a difference, heuristically, and not worth breaking a feature that already exists.

**Comment by ghost** [comment]
paint

Funny drawing to explain the intended cache hits.

File: src/endpoints/backends/chat-completions.js [hunk 2]
```diff
@@ -876,6 +899,47 @@ router.post('/generate', jsonParser, function (request, response) {
         if (request.body.use_fallback) {
             bodyParams['route'] = 'fallback';
         }
+
+        let cachingAtDepth = getConfigValue('claude.cachingAtDepth', -1);
```
Comment (Cohee1207) [inline src/endpoints/backends/chat-completions.js:893]: I'd maybe put that into a separate function (i.e in the prompt-converters module) cause it makes the endpoint harder to read.

Comment (ghost) [reply src/endpoints/backends/chat-completions.js:893]: Done, also edited up the anthropic code so it doesn't rely on ST's current squashing behavior for messages, anything else?

EDIT: by "squashing" I mean "flattening the content arrays and/or just intentionally always putting everything in a single content array".

**Review by Cohee1207 — APPROVED** [review]
I poked at it (with N = 0) and it appears to be functional in the ideal circustances. Can't say any more than that.

**Comment by ghost** [comment]
Thx. Sorry about the eslint thing.

**Pull request merged**